modules/core/src/main/java/org/apache/ignite/internal/processors/query/h2/twostep/messages/GridQueryNextPageResponse.java

```diff
@@ -60,14 +60,18 @@ public class GridQueryNextPageResponse implements Message {
     @GridDirectCollection(Message.class)
     private Collection<Message> vals;
 
-    /** */
+    /**
+     * Note, columns count in plain row can differ from {@link #cols}.
+     * See {@code org.apache.ignite.internal.processors.query.h2.twostep.msg.GridH2ValueMessageFactory#toMessages}.
```

Comment:
Can we also change `GridMapQueryExecutor.prepareNextPage` with this patch?
Now we have this call:
```
toMessages(rows, new ArrayList<>(res.columnCount()), res.columnCount())
```
But actially we should preallocate `res.columnCount() * rows.size()` items, to avoid resizing.

Reply:
Done